core/shared/src/main/scala/fs2/Stream.scala

```diff
@@ -1311,6 +1311,51 @@ final class Stream[+F[_], +O] private (private val free: FreeC[Algebra[Nothing,
     */
   def mask: Stream[F, O] = this.handleErrorWith(_ => Stream.empty)
 
+  /**
+    * Like [[Stream.flatMap]] but interrupts the inner stream when new elements arrive in the outer stream.
+    *
+    * The implementation will try to preserve chunks like [[Stream.merge]].
+    *
+    * Finializers of each inner stream are guaranteed to run before the next inner stream starts.
+    *
+    * When the outer stream stops gracefully, the currently running inner stream will continue to run.
+    *
+    * When an inner stream terminates/interrupts, nothing happens until the next element arrives
+    * in the outer stream(i.e the outer stream holds the stream open during this time or else the
+    * stream terminates)
+    *
+    * When either the inner or outer stream fails, the entire stream fails and the finalizer of the
+	* inner stream runs before the outer one.
+    *
+    */
+  def switchMap[F2[x] >: F[x], O2](f: O => Stream[F2, O2])(
+      implicit F2: Concurrent[F2]): Stream[F2, O2] =
+    Stream.eval(Semaphore[F2](1)).flatMap { guard =>
+      Stream.eval(Ref.of[F2, Option[Deferred[F2, Unit]]](None)).flatMap { haltRef =>
+        Stream.eval(SignallingRef.apply(false)).flatMap { innerFailed =>
+          def runInner(o: O, halt: Deferred[F2, Unit]) =
+            Stream.eval(guard.acquire) >> // guard inner to prevent parallel inner streams
+              f(o)
+                .interruptWhen(halt.get.attempt)
+                .onError { case e => Stream.eval(innerFailed.set(true)) }
```

Comment:
like @pchlupacek said, you can get rid of the `SignallingRef` and just do `++ Stream.eval(guard.release)`

Reply:
If I understand correctly, we can just do `Stream.eval(guard.acquire) >> f(o).interruptWhen(halt.get.attempt) ++ Stream.eval(guard.release)` but how do I get the return type to match `Stream[F2, O2]`?

Reply:
use `Stream.eval_`, which is basically `Stream.eval(..).drain`

Reply:
`drain` gets from `Stream[F, A]` to `Stream[F, Nothing]`, which makes things compile since `Nothing` is a subtype of everything. On a logical level, `drain` evaluates the Stream but doesn't emit anything, hence why `Stream[F, Nothing]`

Reply:
Ah right, perfect, tests are still passing, I'm just gonna check everything again and rebase anyway